### PR TITLE
Frame `remove_all` with size limit.

### LINKED ISSUE #9002: FRAME Storage `remove_all` API should have a key limit
Similar to the `storage_kill` api in child tries, the `remove_all` api on any maps should also have an optional key limit to enforce weight restrictions.

https://github.com/paritytech/substrate/pull/8166

When calling `remove_all`, a user should also be able to include `Option<u32>`, which describes the maximum number of keys that can be removed at once.

At the end of the operation, you can return `AllRemoved(u32)` or `SomeReamining(u32)` with the number of items that were actually removed at the end of the operation.

Should reuse and rename `KillChildStorageResult`:

https://github.com/paritytech/substrate/blob/master/primitives/io/src/lib.rs#L89

### PR DESCRIPTION
**Pull request by cheme**
Fixes #9002.

Note that it breaks existing frame api (could have define new function with the limit but it does not seem good).

Affected APIs:

* `remove_prefix`
* `clear_prefix`
* `kill_prefix`
* `remove_all`

To fix these APIs to act the same as before, simply add an additional input parameter of `None`.

Also rename:

* `KillChildStorageResult` -> `KillStorageResult`

polkadot companion: https://github.com/paritytech/polkadot/pull/3257

### CODE REVIEW

**Comment by shawntabrizi** [comment]
> Note that it breaks existing frame api

The solution is simply to add `None` to the end right? Not actually breaking any logic silently

**Comment by cheme** [comment]
> > Note that it breaks existing frame api
> 
> The solution is simply to add `None` to the end right? Not actually breaking any logic silently

Yes, updating is very straight forward. Seems better to me than having two function for the same thing.

File: primitives/io/src/lib.rs [hunk 1]
```diff
@@ -133,9 +133,42 @@ pub trait Storage {
...
+	///
+	/// Please note that keys that are residing in the overlay for that prefix when
+	/// issuing this call are all deleted without counting towards the `limit`. Only keys
+	/// written during the current block are part of the overlay. Deleting with a `limit`
+	/// mostly makes sense with an empty overlay for that prefix.
+	///
+	/// Calling this function multiple times per block for the same `prefix` does
+	/// not make much sense because it is not cumulative when called inside the same block.
+	/// Use this function to distribute the deletion of a single child trie across multiple
+	/// blocks.
```
Comment (athei) [inline primitives/io/src/lib.rs:161]: Please deduplicate this rather long block of text. I suggest removing all occurrences of it in this file but one and refer to that. Maybe create a heading for that like `# Limit`.

File: primitives/state-machine/src/ext.rs [hunk 2]
```diff
@@ -498,21 +498,46 @@ where
...
+			});
+			(all_deleted, num_deleted)
+		} else {
+			self.backend.apply_to_keys_while(None, Some(prefix), |key| {
+				num_deleted = num_deleted.saturating_add(1);
+				self.overlay.set_storage(key.to_vec(), None);
+				true
+			});
+			(true, num_deleted)
+		}
```
Comment (athei) [inline primitives/state-machine/src/ext.rs:532]: This logic is now copy pasted into two additional functions. We should extract it.

**Review by athei — COMMENTED** [review]
API changes look good to me. But we need to remove some duplication.

**Review by athei — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Comment by thiolliere** [comment]
maybe it worth documenting that the function remove_all, and warn that it removes all overlay, and up to limit in the backend, people shouldn't rely on the fact that it removes exactly up to the limit.

**Pull request merged**